sdk/core/azure-core-experimental/src/main/java/com/azure/core/experimental/util/BinaryData.java

```diff
@@ -0,0 +1,284 @@
+ * array. The data is collected from provided sources and stored into a byte array.
+ * <p>
+ * It also provides a way to serialize and deserialize an {@link Object} into {@link BinaryData} given an
+ * {@link ObjectSerializer}. Code samples are explained below.
+ *
+ * <p><strong>Create an instance from Bytes</strong></p>
+ * {@codesnippet com.azure.core.experimental.util.BinaryDocument.from#bytes}
+ *
+ * <p><strong>Create an instance from String</strong></p>
+ * {@codesnippet com.azure.core.experimental.util.BinaryDocument.from#String}
+ *
+ * <p><strong>Create an instance from InputStream</strong></p>
+ * {@codesnippet com.azure.core.experimental.util.BinaryDocument.from#Stream}
+ *
+ * <p><strong>Get an Object from {@link BinaryData}</strong></p>
+ * {@codesnippet com.azure.core.experimental.util.BinaryDocument.to#ObjectAsync}
+ *
+ * @see ObjectSerializer
+ */
+public final class BinaryData {
+    private final byte[] data;
+
+    BinaryData() {
+        // This exists, so no one is able to create instance, user need to use static function to create instances.
+        this.data = null;
+    }
+
+    /**
+     * Create instance of {@link BinaryData} given the data.
+     * @param data to represent as bytes.
+     */
+    BinaryData(byte[] data) {
+        this.data = data;
+    }
+
+    /**
+     * Provides {@link InputStream} for the data represented by this {@link BinaryData} object.
+     *
+     * <p><strong>Get InputStream from BinaryData</strong></p>
+     * {@codesnippet com.azure.core.experimental.util.BinaryDocument.to#Stream}
+     *
+     * @return {@link InputStream} representing the binary data.
+     */
+    public InputStream toStream() {
+        return new ByteArrayInputStream(this.data);
+    }
+
+    /**
+     * Provides {@link Mono} of {@link InputStream} for the data represented by this {@link BinaryData} object.
+     *
+     * @return {@link InputStream} representation of the {@link BinaryData}.
+     */
+    public Mono<InputStream> toStreamAsync() {
+        return Mono.fromCallable(() -> toStream());
+    }
+
+    /**
+     * Create {@link BinaryData} instance with given {@link InputStream} as source of data. The {@link InputStream} is
+     * not closed by this function.
+     *
+     * <p><strong>Create an instance from InputStream</strong></p>
+     * {@codesnippet com.azure.core.experimental.util.BinaryDocument.from#Stream}
+     *
+     * @param inputStream to read bytes from.
+     * @throws UncheckedIOException If any error in reading from {@link InputStream}.
+     * @throws NullPointerException if {@code inputStream} is null.
+     * @return {@link BinaryData} representing the binary data.
+     */
+    public static BinaryData fromStream(InputStream inputStream) {
+        Objects.requireNonNull(inputStream, "'inputStream' cannot be null.");
+
+        final int bufferSize = 1024;
+        try {
+            ByteArrayOutputStream dataOutputBuffer = new ByteArrayOutputStream();
```

Comment:
Once this moves into `azure-core` we have a specialized `ByteArrayOutputStream` type that allows access to its internal buffer without copying when `toByteArray` is called.

Reply:
yes we can do this when we move it azure-core